### ISSUE TITLE
Performance anomalies on direct transfers.

### ISSUE THREAD

**Issue by dwaddington** [IBM/mcas#81]
get_direct and put_direct performance is low.  It should be closer to 8-9 GB/s

- 4MB value READ (get_direct) performance with hstore: 90 IOPS (271 MB/s)
- 4MB value READ (get_direct) performance with mapstore: 96 IOPS (290 MB/s)

- 4MB value WRITE (put_direct) performance with hstore: 722 IOPS (2166 MB/s)
- 4MB value WRITE (put_direct) performance with mapstore: 2833 IOPS (8501 MB/s)

**kvstore-perf command line**

```
./build/dist/bin/kvstore-perf --core 0 --elements 5000 --value_length 3000000 --test get_direct --server 10.0.0.101 --port 11910 --size 40000000000 --device_name mlx5_0 --component mcas
```

** Raw b/w confirmation **

```
ib_send_bw -s 4000000 -d mlx5_0

************************************
* Waiting for client to connect... *
************************************
Requested SQ size might be too big. Try reducing TX depth and/or inline size.
Current TX depth is 128 and  inline size is 0 .
---------------------------------------------------------------------------------------
                    Send BW Test
 Dual-port       : OFF          Device         : mlx5_0
 Number of qps   : 1            Transport type : IB
 Connection type : RC           Using SRQ      : OFF
 RX depth        : 512
 CQ Moderation   : 100
 Mtu             : 4096[B]
 Link type       : Ethernet
 GID index       : 3
 Max inline data : 0[B]
 rdma_cm QPs     : OFF
 Data ex. method : Ethernet
---------------------------------------------------------------------------------------
 local address: LID 0000 QPN 0x232a PSN 0xd10550
 GID: 00:00:00:00:00:00:00:00:00:00:255:255:10:00:00:101
 remote address: LID 0000 QPN 0x26191 PSN 0x9a71cf
 GID: 00:00:00:00:00:00:00:00:00:00:255:255:10:00:00:124
---------------------------------------------------------------------------------------
 #bytes     #iterations    BW peak[MB/sec]    BW average[MB/sec]   MsgRate[Mpps]
Conflicting CPU frequency values detected: 1000.684000 != 1068.004000. CPU Frequency is not max.
 4000000    1000             0.00               11695.49                   0.003066
---------------------------------------------------------------------------------------


[danielwaddington@sg-mcas104 mcas]$ ib_send_bw -s 4000000 -d mlx5_0 10.0.0.101
Requested SQ size might be too big. Try reducing TX depth and/or inline size.
Current TX depth is 128 and  inline size is 0 .
---------------------------------------------------------------------------------------
                    Send BW Test
 Dual-port       : OFF		Device         : mlx5_0
 Number of qps   : 1		Transport type : IB
 Connection type : RC		Using SRQ      : OFF
 TX depth        : 128
 CQ Moderation   : 100
 Mtu             : 4096[B]
 Link type       : Ethernet
 GID index       : 3
 Max inline data : 0[B]
 rdma_cm QPs	 : OFF
 Data ex. method : Ethernet
---------------------------------------------------------------------------------------
 local address: LID 0000 QPN 0x26191 PSN 0x9a71cf
 GID: 00:00:00:00:00:00:00:00:00:00:255:255:10:00:00:124
 remote address: LID 0000 QPN 0x232a PSN 0xd10550
 GID: 00:00:00:00:00:00:00:00:00:00:255:255:10:00:00:101
---------------------------------------------------------------------------------------
 #bytes     #iterations    BW peak[MB/sec]    BW average[MB/sec]   MsgRate[Mpps]
 4000000    1000             11681.98            11681.97		   0.003062
---------------------------------------------------------------------------------------

```

**Comment by dwaddington** [comment]
What is also strange is that while mapstore put_direct seems somewhere reasonable, mapstore get_direct is x0.25 that of put_direct which is unexpected. hstore swings the other way!

Additional information: 4MB value, one client thread, one shard (6000 pairs, 10 repeats). Using mpi-bench.

IOPS as follows:
**hstore:** get_direct=~400 put_direct=~80
**hstore-cc** : get_direct=~2100 put_direct=~600-900

**Comment by dwaddington** [comment]
Another variable could be that we updated libfabric.

**Comment by dickeycl** [comment]
I seem to remember the standard deviation of RDMA times being quite large. I didn't pursue it, as I didn't think we had any control over those tmes.

**Comment by dwaddington** [comment]
But event the current best case is pretty bad.  The signalling and PM overhead shouldn't be so high.

**Comment by dwaddington** [comment]
Clem, testing with mapstore, I think a central part of the problem is MPI's choice of CPU on the client side. If a core is chosen in a different NUMA zone from the RNIC, then perform drops by about 10x. I'm stilling working on this.
I did run into  problems using 4MB with hstore-cc return E_TOO_BIG on trying to lock the value.

**Comment by dwaddington** [comment]
I can confirm that the READ (get_direct) performance instability was caused by the clients moving across NUMA zones (mapstore=OK, hstore=OK, hstore-cc=EXCEPTION)
However, when the number of clients hitting the shard is increased (e.g. 5 per shard vs. 1 per shard) performance drops dis-proportionally.

Write performance is very low for hstore/hstore-cc. Can we turn off the explicit flushing?

**Comment by dickeycl** [comment]
Explicit flush is specified in Shard::release_locked_value_exclusive: the call to unlock specifies IKVStore::UNLOCK_FLAGS_FLUSH. I will change perf to let an environment variable control it.

**Comment by dickeycl** [comment]
I will look into the E_TOO_BIG trouble. In the past that has usually been a true "out of space" condition. env var CCA_FINE_TRACE=1 will print an allocation map if allocate fails. (The map is also printed in a few success cases.)

**Comment by dwaddington** [comment]
How is the flushing on put_direct different from the memcpy get put ? Since we don't see the performance issue on the memcpy ops?

**Comment by ykorman** [comment]
> For reference: info on DDIO -- https://software.intel.com/content/www/us/en/develop/articles/persistent-memory-replication-over-traditional-rdma-part-1-understanding-remote-persistent.html

I recommend reading the [RPMA](https://github.com/pmem/rpma) docs - https://pmem.io/rpma/
Specifically, this might interest you - https://pmem.io/rpma/documentation/basic-direct-write-to-pmem.html

**Comment by dickeycl** [comment]
MCAS "put direct" uses the sequence described by "General-Purpose Remote Replication Method", but not through librpmem. We do use libpmem for pmem_persist.

**Comment by dickeycl** [comment]
I was able to get a E_TOO_BIG error using hstore-cc with this test:

CCA_FINE_TRACE=1 VALUE_LENGTH=4194304 ELEMENT_COUNT=1256 STORE_SIZE=8000000000 ./dist/testing/mcas-hstore-cc-put_direct-0.sh

The error came after creating 1240 keys and 1239 values, and was caused by the allocator running out of space. Due to the use of binary sizes, Each 4MiB allocation required 4MiB+24 bytes of space, which rounded to 6MiB of space. Sizes from 2MiB to 128MiB round up to a multiple of 2MiB. The waste is actually worse with 2MiB blocks which, including the 24-byte metadata header, round up to 4MiB.

There doesn's seem to be a "table" feature in this log, but the available region characteristics were:
region size | allocation unit
8 GiB | 128 MiB
128 MiB | 2MiB
2 MiB | 32 KiB
32 KiB | 512B
512 B | 8B

Since we didn't have quite 8 GiB, the top region was a bit smaller than 8 GiB.

The 8GB region was eventually split into 59 2^27-byte (128MiB) regions. Each of those regions held 21 6MiB allocations (126MiB) and 2 MiB for overhead and smaller allocations. (59*21=1239). The 8B+24B keys fit into the space for smaller allocations.

In short, the allocator is, in this case, "working as designed." It has been suggested that (a) the 24-byte metadata be placed elsewhere, and not concatenated with the large, possibly 2^n-byte data area, or (b) the allocator impose a smaller rounding requirement than, in this case, 2MiB, or (c) the rounding rule in the allocator somehow acknowledge that 2^n+delta is a common size.

(a) is difficult if for no other reason than deallocation needs to know area size, and only the 24-byte metadata are contains that size. (b) is difficult because finer-grained sizes require more bite to track them, and there is a limit if 64 (or perhaps 128) bits for atomic store operations. I haven't thought much about (c).

**Comment by dickeycl** [comment]
Getting back to "how is memcpy flush different from RDMA write flush", on a single machine (server on core 0, client on core 14), put_direct seems to be better than put. Server invocation is

$ DAX_RESET=1 ./dist/bin/mcas --config '{"shards": [{"core": 0, "addr": "10.0.0.91", "port": 11911, "default_backend": "hstore", "dax_config": [{"addr": 618475290624, "path": "/dev/dax0.0"}]}]}' --forced-exit

Client invocation is./dist/bin/kvstore-perf --cores 14 --src_addr 10.0.0.91 --server 10.0.0.91 --test put_direct --component mcas --elements 3000 --size 18000048000 --skip_json_reporting --key_length 8 --value_length 2000000

 [TOTAL] 14 put_direct IOPS: 1079 

Changing "put_direct" to "put" in the client,

 [TOTAL] 14 put IOPS: 778

If put_direct becomes slower I presume that happens due to multiple clients, or remote clients, or both.

**Comment by dickeycl** [comment]
Repeated tests with same-size (2MB) get, get_direct, put and put_direct showed that some runs would fail to start, and others would report about 50 IOps rather than ~1K IOps. The complete failures were due to failure to allocate a "receive" buffer before connection. (The recipient of the first RDMA transfer would accept a connection and then allocate a receive buffer. If the first RDMA arrived before the receive buffer was allocated, that RDMA would fail.)

After fixing the RDMA race condition tests never fail to start, and I cannot replicate the 50 IOps behavior. That is, runs are consistently on the order of K IOps.

In 800 runs with 8-byte key, 2MB data, min to max measurements were:
put: 794 .. 835 IOps
get: 2071 .. 2113 IOps
put_direct: 1295 .. 1355 IOps
get_direct: 2659 .. 2725 IOps

My guess is that the initial race condition managed to affect some runs which did not fail entirely, leading to the low (50 IOps) numbers.

**Comment by dwaddington** [comment]
@dickeycl here is the latest release regression-run performance (machine kivati2, devdax).  This still feels to me to be under-performing for large 2MB write operations.
```
Test mcas-mapstore-put-8-8: passed (173679 of 170000 IOPS)
Test mcas-mapstore-get-8-8: passed (207792 of 180000 IOPS)
Test mcas-hstore-put-8-8-devdax: passed (191273 of 185000 IOPS)
Test mcas-hstore-get-8-8-devdax: passed (214451 of 200000 IOPS)
Test mcas-hstore-put-8-2000000-devdax: passed (384 of 350 IOPS)
Test mcas-hstore-get-8-2000000-devdax: passed (1845 of 1800 IOPS)
Test mcas-hstore-put_direct-8-2000000-devdax: fail (544 of 1250 IOPS)
Test mcas-hstore-get_direct-8-2000000-devdax: passed (3457 of 1250 IOPS)
Test mcas-hstore-cc-kvtest-0-devdax: passed
Test mcas-hstore-cc-put-8-8-devdax: fail (177656 of 180000 IOPS)
Test mcas-hstore-cc-get-8-8-devdax: passed (215870 of 180000 IOPS)
Test mcas-hstore-cc-put-8-2000000-devdax: fail (388 of 750 IOPS)
Test mcas-hstore-cc-get-8-2000000-devdax: passed (1837 of 1800 IOPS)
Test mcas-hstore-cc-put_direct-8-2000000-devdax: fail (561 of 1250 IOPS)
Test mcas-hstore-cc-get_direct-8-2000000-devdax: passed (3473 of 1250 IOPS)
Test mcas-hstore-dax-conflict-0-devdax: passed

```

**Comment by dwaddington** [comment]
With export FLUSH_ENABLE=0 (see below). So it appears to be cache flushing cost. Let's look at the flushing code. (BTW, there is a but on the 750 threshold).

Test mcas-mapstore-put-8-8: passed (176412 of 170000 IOPS)
Test mcas-mapstore-get-8-8: passed (213155 of 180000 IOPS)
Test mcas-hstore-put-8-8-devdax: passed (192449 of 185000 IOPS)
Test mcas-hstore-get-8-8-devdax: passed (218705 of 200000 IOPS)
Test mcas-hstore-put-8-2000000-devdax: passed (384 of 350 IOPS)
Test mcas-hstore-get-8-2000000-devdax: passed (1841 of 1800 IOPS)
Test mcas-hstore-put_direct-8-2000000-devdax: passed (3164 of 1250 IOPS)
Test mcas-hstore-get_direct-8-2000000-devdax: passed (3465 of 1250 IOPS)
Test mcas-hstore-cc-kvtest-0-devdax: passed
Test mcas-hstore-cc-put-8-8-devdax: passed (181517 of 180000 IOPS)
Test mcas-hstore-cc-get-8-8-devdax: passed (215015 of 180000 IOPS)
Test mcas-hstore-cc-put-8-2000000-devdax: fail (388 of 750 IOPS)
Test mcas-hstore-cc-get-8-2000000-devdax: passed (1857 of 1800 IOPS)
Test mcas-hstore-cc-put_direct-8-2000000-devdax: passed (3209 of 1250 IOPS)
Test mcas-hstore-cc-get_direct-8-2000000-devdax: passed (3467 of 1250 IOPS)
Test mcas-hstore-dax-conflict-0-devdax: passed

**Comment by dickeycl** [comment]
The "flushing code" may cause the entire RDMAed region to be flushed. Techniques to make it go faster are discussed in the Intel papers cited above.
